pathchk implemented (see #841)

ibabushkin opened the pull request:
This is my promised implementation of `pathchk`. Should be bug-free, hopefully ideomatic as well.
Code is comented, as I am new to the language.

In the meantime, I realized someone claimed this issue ~2 years back, but apparently didn't do any work since then.

Heather commented:
@ebfe @jbcrail can someone check / review this?

Arcterus commented on src/pathchk/pathchk.rs:
```diff
@@ -0,0 +1,229 @@
+static NAME: &'static str = "pathchk";
+static VERSION: &'static str = env!("CARGO_PKG_VERSION");
+
+// a few global constants as used in the GNU implememntation
```
Did you pull these constants from the GNU source code or did you get them some other way?

ibabushkin replied:
The modes are essentially abstract representations of the commandline flags and specify the behaviour of the utility.

Arcterus commented on src/pathchk/pathchk.rs:
```diff
@@ -0,0 +1,229 @@
+        _ => {
+            let mut res = true;
+            if matches.free.len() == 0 {
+                writeln!(&mut std::io::stderr(),
```
Probably use `show_error!` here instead.

ibabushkin replied:
On line 79, I assume?

ebfe commented:
Please add some basic tests. In the Makefile add it to either PROGS or UNIX_PROGS and not both. UNIX_PROGS is for utils that work on unix only.

ibabushkin commented:
Thanks for your feedback. I will try to write tests and incorporate your suggestions as soon as possible.

ibabushkin commented:
I incorporated the suggestions by @Arcterus and fixed the makefile. Also, I added the first test for the utility, but I am unsure whether it is compliant with the project (from what I've seen most tests look similar, but the quality standards might bite me here). More tests about to follow.

Heather commented:
tests/pathchk.rs:21:34: 21:44 error: unresolved name `repeat_str` [E0425]
(there are still being travis issues)

ibabushkin commented:
This is very weird: `repeat_str()` is defined in `tests/common/util.rs`:
```Rust
pub fn repeat_str(s: &str, n: u32) -> String {
    let mut repeated = String::new();
    for _ in 0..n {
        repeated.push_str(s);
    }
    repeated
}
```
On my machine, using nightly Rust, everything compiles and runs fine:
```
$ make UTILS=pathchk test
cargo build   -p pathchk
cargo test  --test pathchk --features "pathchk " --no-default-features 
   Compiling uutils v0.0.1 (file:///home/thewormkill/clones/forks/coreutils)
     Running target/debug/pathchk-a5ab1a66f4c7bd98

running 1 test
test test_default_mode ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
```
I don't know what could cause this *at all* ;).

ibabushkin commented:
Okay, apparently AppVeyor fails with this (after a lot of hickups):
```
cargo build  --features "base64 basename cat cksum comm cp cut dirname echo env expand expr factor false fmt fold hashsum head link ln ls mkdir mktemp nl nproc od paste printenv printf ptx pwd readlink realpath relpath rm rmdir seq shred shuf sleep sort split sum sync tac tail tee test tr true truncate tsort unexpand uniq wc whoami yes" --release --no-default-features
/bin/sh: cargo: command not found
Makefile:242: recipe for target 'build-uutils' failed
make: *** [build-uutils] Error 127
Command exited with code 2
```
Travis essentially found a failure I introduced by pushing to or from the wrong branch at some point (see commit message).

ibabushkin commented:
Now, AppVeyor is failing for some other reason not directly related to the code (it doesn't even start compiling).

Heather commented:
@ibabushkin yes, appVeyor is broken currently, may you squash not imformative commits else I can merge with squash this request

ibabushkin commented:
Which exactly should I squash? My git fuckup leading to this becoming so convoluted set me of-track for a biti, sorry :D

Heather commented:
@ibabushkin merge commits

ibabushkin commented:
Hm, I could just rebase everything, but because it is pushed to the repo already, I might have to create a new PR. is that alright?

Heather commented:
@ibabushkin depends... you can push to new branch and open new PR to be safe or you push with force and then this PR is still being actual

ibabushkin commented:
Oh, seems like I misunderstood you. Of course you may squash my commits. Sorry for the misunderstandings I caused, seeems I really need more sleep.
